chroma-core/chroma
# OTel tracing throughout the codebase

## Description of changes

This PR adds OpenTelemetry tracing to ~all major methods throughout our codebase. It also adds configuration to specify where these traces should be sent. I focused more on laying the groundwork for tracing than on collecting all the data we need everywhere.

Default behavior is unchanged: no tracing, no printing.

*Summarize the changes made by this PR.*
 - New functionality
	 - OpenTelemetryClient with relevant config.
	 - Wrap most of our code in tracing.

I made two design decisions as part of this PR:
- `with` statements instead of decorators
- Fully separate OpenTelemetry stuff and Posthog (product telemetry) stuff.

My philosophy with these decisions, as with most of the software I write, is that we should err on the side of being slightly verbose and explicit over terse and implicit.

### `with` statements instead of decorators

At first I thought we could do something like [OpenTelemetry's trace decorators](https://opentelemetry.io/docs/instrumentation/python/manual/). Decorators like this have two big problems for our use-case:
- They destroy type information. `functools.wraps` preserves a small amount (just enough that `@override` still works) but, notably, it still clobbers information about parameters and return values. In other words, using a tracing decorator would remove type information from all the wrapped functions. Support for decorator type hinting [arrived in Python 3.10](https://peps.python.org/pep-0612/) which means we can eventually switch. [1]
- A tracing decorator is necessarily very general, recording the same data for every single function. As time goes on we will want to add specific data to spans depending on the code they wrap; for example, we'll want to add `collection_id`, `vectors_count`, and others to `collection.add()` traces even though they don't make sense in other contexts. This means that even if we used a decorator to preserve readability, we would still need to add attributes to the current span within method bodies.

Because of these two considerations I don't think a tracing decorator is worth it right now: it would effectively make our codebase untyped without even significantly increasing readability.

[1] I hoped we could do the type checking in Python>=3.10 since it's only for development and not at runtime, but doing something like [this StackOverflow code](https://stackoverflow.com/a/68290080) (unsurprisingly) refuses to run with `python3.9`.

### Separate OTel and product telemetry

It's tempting to combine OTel and product telemetry behind a single internal interface. I don't think this coupling is worth it. Product telemetry cares about a small and relatively static set of uses, whereas tracing by nature should be very deep in our codebase. I see two ways to couple them and problems with each:
- Have a unified Telemetry interface only for the events our product telemetry cares about and use raw OTel for the rest. In other words, use this telemetry interface only for `collection.add()`s, `collection.delete()`s, etc. This seems weird to me: tracing code would be implicit in some cases but explicit in others, making the codebase less easily comprehensible. Also if an engineer later decides to add product telemetry to a codepath that already has tracing, they need to know to remove existing tracing. This increases the cognitive overhead required to work on Chroma, reducing the readability and maintainability of our codebase.
- Have a unified Telemetry interface which does everything. In this case, it has the above behavior but also wraps all other OTel behavior we want. This seems weird to me because we're basically writing a wrapper around the complete set of OpenTelemetry behavior which only modifies a small part of it. This increases our maintenance burden for very little value.


## Test plan
*How are these changes tested?*

- [ ] Tests pass locally with `pytest` for python, `yarn test` for js

Manual testing:
- Set environment variables to export traces to Honeycomb at various granularities.
- Went through various basic Chroma flows and checked that traces show up in Honeycomb as expected.

## Documentation Changes
*Are all docstrings for user-facing APIs updated if required? Do we need to make documentation changes in the [docs repository](https://github.com/chroma-core/docs)?*

Docs PR to land before this does.
